core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -96,57 +97,55 @@ class ReplicaAlterLogDirsThread(name: String,
+      val PromotionState(reassignmentState, topicId, originalDir) = this.promotionState(topicPartition)
+      if (reassignmentState.inconsistentMetadata) {
+        directoryEventHandler.handleAssignment(new TopicIdPartition(topicId, topicPartition.partition()), originalDir, () => ())
+      }
+
+      this.promotionStates.remove(topicPartition)
+    }
 
     super.removePartitions(topicPartitions)
   }
 
+  private def promotionState(topicPartition: TopicPartition): PromotionState = promotionStates.get(topicPartition)
+
+  private def reassignmentState(topicPartition: TopicPartition): ReassignmentState = promotionState(topicPartition).reassignmentState
+
   // Visible for testing
-  private[server] def updatedAssignmentRequestState(topicPartition: TopicPartition)(state: ReplicaAlterLogDirsThread.DirectoryEventRequestState): Unit = {
-    assignmentRequestStates.put(topicPartition, state)
+  private[server] def updateReassignmentState(topicPartition: TopicPartition, state: ReassignmentState): Unit = {
+    promotionStates.put(topicPartition, promotionState(topicPartition).withAssignment(state))
   }
 
   private def maybePromoteFutureReplica(topicPartition: TopicPartition, partition: Partition) = {
     val topicId = partition.topicId
     if (topicId.isEmpty)
       throw new IllegalStateException(s"Topic ${topicPartition.topic()} does not have an ID.")
 
-    partitionAssignmentRequestState(topicPartition) match {
-      case None =>
+    reassignmentState(topicPartition) match {
+      case ReassignmentState.None =>
         // Schedule assignment request and don't promote the future replica yet until the controller has accepted the request.
         partition.runCallbackIfFutureReplicaCaughtUp(_ => {
-          partition.futureReplicaDirectoryId()
-            .map(id => {
-              directoryEventHandler.handleAssignment(new TopicIdPartition(topicId.get, topicPartition.partition()), id,
-                () => updatedAssignmentRequestState(topicPartition)(ReplicaAlterLogDirsThread.COMPLETED))
-              // mark the assignment request state as queued.
-              updatedAssignmentRequestState(topicPartition)(ReplicaAlterLogDirsThread.QUEUED)
-            })
+          val targetDir = partition.futureReplicaDirectoryId().get
+          val topicIdPartition = new TopicIdPartition(topicId.get, topicPartition.partition())
+          directoryEventHandler.handleAssignment(topicIdPartition, targetDir, () => updateReassignmentState(topicPartition, ReassignmentState.Accepted))
+          updateReassignmentState(topicPartition, ReassignmentState.Queued)
         })
-      case Some(ReplicaAlterLogDirsThread.COMPLETED) =>
+      case ReassignmentState.Accepted =>
         // Promote future replica if controller accepted the request and the replica caught-up with the original log.
         if (partition.maybeReplaceCurrentWithFutureReplica()) {
```

Comment:
Should we log something in debug level or trace level, if `maybeReplaceCurrentWithFutureReplica` return false, because of not caught-up?

Reply:
Same as with L127, let's continue in the other thread.